### PR TITLE
[LLVM] Make more use of IRBuilder::CreateIntrinsic. NFC.

### PR DESCRIPTION
**Pull request by jayfoad**
Convert almost every instance of:
  CreateCall(Intrinsic::getOrInsertDeclaration(...), ...)
to the equivalent CreateIntrinsic call.

### CODE REVIEW

**Comment by jayfoad** [comment]
I could not convert this instance because it uses operand bundles, which are not supported by `CreateIntrinsic`: https://github.com/llvm/llvm-project/blob/87f126243beb69b8b02e5cd4df762bc8a6f1f8cc/llvm/lib/Transforms/Instrumentation/PGOInstrumentation.cpp#L1040

File: llvm/lib/IR/AutoUpgrade.cpp [hunk 1]
```diff
@@ -2325,25 +2323,21 @@ static Value *upgradeNVVMIntrinsicCall(StringRef Name, CallBase *CI,
...
-        Arg, "ctpop");
+    Value *Popc = Builder.CreateIntrinsic(Intrinsic::ctpop, {Arg->getType()},
+                                          Arg, /*FMFSource=*/nullptr, "ctpop");
     Rep = Builder.CreateTrunc(Popc, Builder.getInt32Ty(), "ctpop.trunc");
   } else if (Name == "h2f") {
-    Rep = Builder.CreateCall(Intrinsic::getOrInsertDeclaration(
-                                 F->getParent(), Intrinsic::convert_from_fp16,
-                                 {Builder.getFloatTy()}),
-                             CI->getArgOperand(0), "h2f");
+    Rep = Builder.CreateIntrinsic(Intrinsic::convert_from_fp16,
```
Comment (arsenm) [inline llvm/lib/IR/AutoUpgrade.cpp:2338]: We should stop using these intrinsics and just use fpext

**Review by arsenm — APPROVED** [review]

File: llvm/lib/CodeGen/StackProtector.cpp [hunk 2]
```diff
@@ -541,8 +540,7 @@ static bool CreatePrologue(Function *F, Module *M, Instruction *CheckLoc,
   AI = B.CreateAlloca(PtrTy, nullptr, "StackGuardSlot");
 
   Value *GuardSlot = getStackGuard(TLI, M, B, &SupportsSelectionDAGSP);
-  B.CreateCall(Intrinsic::getOrInsertDeclaration(M, Intrinsic::stackprotector),
-               {GuardSlot, AI});
+  B.CreateIntrinsic(Intrinsic::stackprotector, {}, {GuardSlot, AI});
```
Comment (jurahul) [inline llvm/lib/CodeGen/StackProtector.cpp:543]: Looks like we could use an overload where we do not need to specify Types or RetTy (for non-overloaded intrinsics).

**Review by jurahul — APPROVED** [review]

File: llvm/lib/CodeGen/StackProtector.cpp (hunk 2 above)
Comment (jayfoad) [reply llvm/lib/CodeGen/StackProtector.cpp:543]: Yeah, there could certainly be more useful overloads, and maybe more default arguments. But I didn't really want to do that in this patch.

**Comment by jurahul** [comment]
Looks like this is causing some build failures due to unused variables:

/home/buildbots/llvm-external-buildbots/workers/ppc64le-lld-multistage-test/ppc64le-lld-multistage-test/llvm-project/llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp:1868:13: error: unused variable 'M' [-Werror,-Wunused-variable]
 1868 |     Module *M = IRB.GetInsertBlock()->getParent()->getParent();
      |             ^
1 error generated.

/home/buildbots/llvm-external-buildbots/workers/ppc64le-lld-multistage-test/ppc64le-lld-multistage-test/llvm-project/llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1028:11: error: unused variable 'M' [-Werror,-Wunused-variable]
 1028 |   Module *M = IRB.GetInsertBlock()->getParent()->getParent();
      |           ^

**Comment by jayfoad** [comment]
> Looks like this is causing some build failures due to unused variables:

Fixed with 9255850e89b1e538e11fcc8b71cfd0b320546a75. (But since when have GCC's unused variable warnings been treated as build errors?)

**Pull request merged**